telemetry/core/src/main/scala/io/buoyant/telemetry/KafkaTelemeter.scala

```diff
@@ -0,0 +1,205 @@
+    new KafkaTelemeter("zipkin", numRetries, sampleRate, brokerList, metrics, tracing)
+  }
+}
+
+//Going with hybrid approach. Let kafka handle retries. Batching will be handled outside as it is being done today.
+//We will leave linger.ms unchanged for this reason at 1 ms.
+//The reason for this is that the API has been constructed in such a way that a given trace is annotated across multiple
+//different calls rather than gathering up all its annotations and then sending it on its way.
+
+case class KafkaTelemeter(topic: String, numRetries: Int, sampleRate: Float, brokerList: String, metrics: Boolean, tracing: Boolean) extends Telemeter {
+  private val log = Logger.get(getClass)
+  log.info("Initializing kafka telemeter")
+  private var producer: KafkaProducer[Array[Byte], Array[Byte]] = null
+  private val METADATA_FETCH_TIMEOUT_CONFIG: Int = 3000
+  private val TIMEOUT_CONFIG: Int = 300
+  private val MAX_BLOCK_MS: Int = 1000
+  private val RETRY_BACKOFF_MS_CONFIG: Int = 10000
+  private val RECONNECT_BACKOFF_MS_CONFIG: Int = 10000
+  private val CACHE_BYTE_ARRAYS: Int = 100
+  //TODO need to do the tracerCache stuff to prevent duplicates in the config as well as allow support for multiple
+  //kafka clusters. TODO track all the different instances of tracers/dedup them.
+
+  def tracer: Tracer = {
+    val tracerImpl = KafkaRawZipkinTracer.tracerCache.putIfAbsent(
+      brokerList + topic,
+      new KafkaTracer(new KafkaRawZipkinTracer(brokerList, numRetries, topic), sampleRate)
+    )
+    KafkaRawZipkinTracer.tracerCache.get(brokerList + topic)
+  }
+
+  def cleanup(): Unit = {
+    log.info("Starting cleanup of the kafka telemeter")
+    //This is a stop gap arrangement. Give one minute for all remaining spans to be flushed.
+    Thread.sleep(1 * 60 * 1000)
+    //Then wait for graceful shutdown of the kafka producer to ensure things have been emitted.
+    producer.close()
+    log.info("Done cleaning up the kafka telemeter")
+  }
+
+  object KafkaRawZipkinTracer {
+    val tracerCache = new ConcurrentHashMap[String, KafkaTracer]
+  }
+
+  class KafkaTracer(tracer: Tracer, sampleRate: Float = Sampler.DefaultSampleRate)
+    extends SamplingTracer(tracer, sampleRate)
+
+  //Only support thrift for now.
+  class KafkaRawZipkinTracer(
+    brokerList: String,
+    numRetries: Int = 3, //this is the kafka default.
+    topic: String = "zipkin",
+    timer: Timer = DefaultTimer.twitter,
+    statsReceiver: StatsReceiver = DefaultStatsReceiver
+  ) extends RawZipkinTracer(statsReceiver, timer) {
+    log.info("Initializing kafka zipkin tracer")
+    private var producer: KafkaProducer[Array[Byte], Array[Byte]] = null
+    val configs: java.util.Map[String, AnyRef] = new java.util.HashMap[String, AnyRef]
+    configs.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, brokerList)
+    configs.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
```

Comment:
I want to perform minimal changes to the telemeter framework in order to get this to work. This will enable us to get updates from linkerd without having to worry about rebasing issues. (Currently I have made a single line change to an existing linkerd file).